### ISSUE TITLE
Performance issue

### ISSUE THREAD

**Issue by tibotiber** [williamngan/react-pts-canvas#4]
Hi @williamngan,

I've had some performance issue in my app which I narrowed down to the `PtsCanvas` component. To be precise, when this component gets re-rendered many times (e.g. from prop change, the CPU usage goes up to more than 90%).

I've checked and it seems to come from the call to `play()` L25 below: 
https://github.com/williamngan/react-pts-canvas/blob/6e3db181bd755129d7da65eec4055f81c4e667fc/src/index.js#L19-L34

I'm guessing `play()` should only be called if there was a change in props, not every time. Another way would be to make sure `play` doesn't cause this kind of performance issue in the ptsjs lib.

I can try to open a PR, but would rather get your opinion on this first.
✌️

**Comment by williamngan** [comment]
Wow, thanks for catching this @tibotiber . CPU 90% is not good!

I wonder if we should try useEffect or some newer React magic. I might be quite busy in the next few weeks so not sure if I can look into this quickly. (Also my React knowledge is so-so 😓)

I guess we need to find out if this is a React implementation issue or Pts issue first. It would be simpler if we can fix this by a different React implementation.

If you come across an idea to fix this, please definitely open a PR! 🙏

**Comment by tibotiber** [comment]
`useEffect` sounds like a good idea, so it will only react to changes in the `play` value. I can try to have a go at this like that.

The alternative I had in mind was that calling `space.play()` multiple times will not request multiple animation frames (see [here](https://github.com/williamngan/pts/blob/master/src/Space.ts#L97)). That would be a fix directly in the Pts lib. But I'm not too sure how to do this, I don't really understand the way this method is called recursively. 

What would you rather?

**Comment by tibotiber** [comment]
@williamngan I realize i might assume too much you understand what I'm speaking about, let me correct that. The perf bug comes from `space.play()` being called many time. From what I understand of [the code](https://github.com/williamngan/pts/blob/master/src/Space.ts#L97), I'm assuming the issue comes from making multiple calls to `requestAnimationFrame` which are never cancelled. So I wanted to track somehow whether the call has already been made. I had a go at this, but couldn't really figure how that method is recursively called (as noted in your comments).

My feeling is we should fix that there, in ptsjs. If that too complicated, then we can play with props diffing or `useEffect` here, and I'm fairly confident I can figure it out without further input. So your call on which side you prefer we work. In both case, I can do the heavy lifting, but in the first I might need some pointers.

Also to keep in mind, if we fix things here, `useEffect` would mean we don't support react under 16 (iirc?). Is that ok for you, or would you rather good old props comparison and support for older versions of react? Not sure what your dependents typically use.

**Comment by williamngan** [comment]
Ah I see. Agree with you that this looks like a Pts issue.

It might be a simple fix - we just need to check if `_animID` has been set already, and if so, call `cancelAnimationFrame` before starting new ones, right? 

The 'recursively' in my comment is probably not the right wording. What I meant (I think 😛) is that we the requestAnimationFrame will call the `play` function again.

Let me know what you think. I can also try a fix this weekend.

**Comment by tibotiber** [comment]
ok, I had missed that `play` was the callback of the animation frame 😅. I'm understanding it better now. I've pushed a PR fixing this. Tested it in my app and seems good. 

Basically, `_animID` is usually set since `play` is recursively called, but I make sure to unset it if we cancel the animation frame and when play is called "manually" (i.e. `time = 0`), I make sure it's not set, or I bail out of starting a new play loop.
